pyproject: Drop python 3.6

tony opened the pull request:
Resolves: https://github.com/python-poetry/poetry/issues/4997

# Pull Request Check List

- [ ] Requires https://github.com/python-poetry/poetry-core/pull/263
- [ ] See also: https://github.com/python-poetry/install.python-poetry.org/pull/11

Resolves: #4997

- [ ] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

neersighted commented:
There are a couple more things to do here, I think: 

* the Python version in `.flake8`
* the arguments to pyupgrade in `.pre-commit-config.yaml`

tony commented:
> There are a couple more things to do here, I think:
> 
> * the Python version in `.flake8`
> * the arguments to pyupgrade in `.pre-commit-config.yaml`

I will convert it to a draft, handle those things, and also continue to look for any other places

If you notice anything else let me know

tony commented:
pre-commit.ci autofix

neersighted commented:
Looks like we can drop the `dataclasses` dep as the marker (`python = "<3.7"`) makes it a no-op.

tony commented:
@neersighted 

- Dropped `dataclasses`
  - poetry-core: Also already dropped `dataclasses` in https://github.com/python-poetry/poetry-core/pull/263
- [`types-dataclasses`](https://pypi.org/project/types-dataclasses/): removed

branchvincent commented on pyproject.toml:
```diff
@@ -60,7 +59,7 @@ pytest = "^6.2"
 pytest-mock = "^3.5"
 pytest-sugar = "^0.9"
-pre-commit = { version = "^2.6", python = "^3.6.1" }
+pre-commit = { version = "^2.6", python = "^3.7" }
```
the `python` marker is no longer necessary
```suggestion
pre-commit = "^2.6"
```

branchvincent commented on tests/installation/test_executor.py:
```diff
@@ -111,7 +111,7 @@ def test_execute_executes_a_batch_of_operations(
 ):
     pip_editable_install = mocker.patch(
-        "poetry.installation.executor.pip_editable_install", unsafe=not PY36
+        "poetry.installation.executor.pip_editable_install", unsafe=not PY37
```
this should now always evaluate to `False`, but since that's the [default](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock) i think we can remove it:
```suggestion
        "poetry.installation.executor.pip_editable_install"
```

tony commented:
@branchvincent: Thank you

Rebased.

Can you approve running on https://github.com/python-poetry/poetry-core/pull/263?

tony commented:
@branchvincent 

P.S. 
![image](https://user-images.githubusercontent.com/26336/150648731-69d99030-8766-4c2a-993d-1c296b4014bb.png)

Any preference on how to navigate this? I can merge the commits into mine or re-do them myself.

branchvincent commented:
> Can you approve running on python-poetry/poetry-core#263?

Unfortunately I don't have access yet 😓 , I've just started helping here. I'll see if I can find someone who does

cc @finswimmer?

> Any preference on how to navigate this?

Thanks for checking, but it's fine as is! I only have it as a requirement for another project

I believe there's one more version-gated block we can remove:

https://github.com/python-poetry/poetry/blob/28d05d002475f0cc55e7cc46ae01f9a3ce4e03a8/src/poetry/utils/pip.py#L58-L69

branchvincent reviewed (approved):
Thanks @tony, this looks good to me. Let's get one more reviewer to approve/merge

tony commented:
> Thanks @tony, this looks good to me. Let's get one more reviewer to approve/merge

Thank you for the review!
